src/indexer.rs

```diff
@@ -132,32 +134,84 @@ async fn sync_state_to_db(
+                sync_state_to_db(abi_item.name.to_lowercase(), csv_writer, db_writers).await;
             }
 
-            if !indexer_config.postgres.apply_indexes_before_sync && !reached_head {
-                println!(
-                    "applying indexes for {}, may take a little while...",
-                    abi_item.name
-                );
-                generate_event_table_indexes(postgres_db, abi_item, &abi_item.name)
-                    .await
-                    .unwrap();
+            // The block below is very specific to postgres
+            // This is some scaffolding to allow this to work for now
+            // TODO: move index creation to postgres implementation + make part of interface for datasource
+            if let Some(postgres_conf) = &indexer_config.postgres {
+                if !postgres_conf.apply_indexes_before_sync && !reached_head {
+                    println!(
+                        "applying indexes for {}, may take a little while...",
+                        abi_item.name
+                    );
+
+                    // Apply to postgres client from db writer (db client) list
+                    for datasource in db_writers {
+                        if let Some(postgres_db) =
+                            datasource.as_any().downcast_ref::<PostgresClient>()
+                        {
+                            generate_event_table_indexes(postgres_db, abi_item, &abi_item.name)
+                                .await
+                                .unwrap();
+                        }
+                    }
+                }
             }
         }
     }
 }
 
+///
+///  This is a factory to declare the various supported datasources
+///  For now this will check existence of each type block to add to implementation (postgres or
+///  bigquery) - but should consider a more generic implementation in the future
+///
+///  # Arguments
+///
+///  * indexer_config: the full configuration    
+pub async fn init_datasource_writers(
+    indexer_config: &IndexerConfig,
+) -> Vec<Box<dyn DatasourceWritable>> {
+    // Return a list / vector of datasource writers
+    let mut writers: Vec<Box<dyn DatasourceWritable>> = Vec::new();
+
+    // Init postgres client (if exists)
+    if let Some(postgres_conf) = &indexer_config.postgres {
+        let postgres_db_client =
+            init_postgres_db(&postgres_conf, &indexer_config.event_mappings, false)
+                .await
+                .expect("Failed to initialize Postgres client");
+        writers.push(Box::new(postgres_db_client));
+    }
+
+    // Init GCP bigquery client (if exists)
+    if let Some(bigquery_conf) = &indexer_config.gcp_bigquery {
+        let bigquery_db_client =
+            init_gcp_bigquery_db(&bigquery_conf, &indexer_config.event_mappings)
+                .await
+                .expect("Failed to initialize bigquery client");
+        writers.push(Box::new(bigquery_db_client));
+    }
+
+    if writers.len() < 1 {
```

Comment:
can these work together at same time though? 🤔 if not then it should just be equal 1 right?

Reply:
i've been running them together (i.e. writign to both local postgres and to bigquery).  if someone wants to write to mulitple datasources (maybe a lower probability scenario), should they be able to do so?

Reply:
i dont know if it makes more sense for them to run to reth-indexers but i guess let them decide this way keeps it open for them